lib/internal/streams/readable.js

```diff
@@ -1123,6 +1124,10 @@ async function* createAsyncIterator(stream, options) {
       next.call(this);
     })
     .on('close', function() {
+      if (endEmitted) {
+        throw new ERR_STREAM_PREMATURE_CLOSE();
```

Comment:
Throwing from an async event handler is not a good idea because it can't be caught (aside from the process-wide `'uncaughtException'` event handler).

Reply:
Ah, I almost forgot about that behavior in js.

Sorry, I am new to the project, so I am unsure about my solution, but what do you think about the following idea?
```javascript
on('close', function() {
  if (endEmitted) {
    stream.emit('error', new ERR_STREAM_PREMATURE_CLOSE());
  } else {
    closeEmitted = true;
    next.call(this);
  }
});
```

I don't quite understand how, but my guess is that `next.call(this)` awakens the while loop that awaits a Promise:

```javascript
else {
  await new Promise(next);
}
```

So my idea is, to emit `error` event with an instance of `ERR_STREAM_PREMATURE_CLOSE` so that it could, in turn, throw the error properly inside the try/catch block:
```javascript
else if (errorEmitted) {
  throw error;
}
```

Why I am unsure:
1. If I emit `error` event, when will it be processed? Will it be the next loop of the event loop?
2. If so, can other events getting processed before the `error` event result in an unwanted behavior?

I believe that I also need an input from @ronag on this

Reply:
You don't want to do anything in the event handler here. You should probably look at the finally block.